dotnet/Gherkin/AstNode.cs

```diff
@@ -11,86 +15,90 @@ public Token GetToken(TokenType tokenType)
         return GetSingle<Token>((RuleType)tokenType);
     }
 
-    public IEnumerable<Token> GetTokens(TokenType tokenType)
+    public Enumerable<Token> GetTokens(TokenType tokenType)
     {
         return GetItems<Token>((RuleType)tokenType);
     }
 
     public T GetSingle<T>(RuleType ruleType)
     {
-        if (!subItems.TryGetValue(ruleType, out var items))
-            return default;
-        if (items is List<object> list)
+        bool foundOne = false;
+        T ret = default;
+        foreach ((var itemType, var item) in subItems)
         {
-            T ret = default;
-            bool foundOne = false;
-            foreach (var item in list)
-            {
-                if (item is T tItem)
-                {
-                    if (foundOne)
-                        throw new InvalidOperationException();
-                    ret = tItem;
-                    foundOne = true;
-                }
-            }
+            if (itemType != ruleType)
+                continue;
+            if (item is not T tItem)
+                continue;
             if (foundOne)
-                return ret;
-            else
                 throw new InvalidOperationException();
+            ret = tItem;
+            foundOne = true;
         }
-        else if (items is T tItem)
-        {
-            return tItem;
-        }
-        return default;
+        return ret;
     }
 
-    public IEnumerable<T> GetItems<T>(RuleType ruleType)
+    public readonly struct Enumerable<T> : IEnumerable<T>
     {
-        if (!subItems.TryGetValue(ruleType, out var items))
-            yield break;
-        if (items is List<object> list)
-        {
-            foreach (var item in list)
-            {
-                if (item is T tItem)
-                    yield return tItem;
-            }
-        }
-        else if (items is T tItem)
+        readonly List<(RuleType RuleType, object Item)> subItems;
+        readonly RuleType ruleType;
+
+        internal Enumerable(List<(RuleType RuleType, object Item)> subItems, RuleType ruleType)
         {
-            yield return tItem;
+            this.subItems = subItems;
+            this.ruleType = ruleType;
         }
-    }
 
-    public void SetSingle<T>(RuleType ruleType, T value)
-    {
-        subItems[ruleType] = new object[] { value };
-    }
+        public Enumerator<T> GetEnumerator() => new Enumerator<T>(subItems.GetEnumerator(), ruleType);
 
-    public void AddRange<T>(RuleType ruleType, IEnumerable<T> values)
-    {
-        foreach (var value in values)
-        {
-            Add(ruleType, value);
-        }
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
     }
 
-    public void Add<T>(RuleType ruleType, T obj)
+    public struct Enumerator<T> : IEnumerator<T>

```

Comment:
Do we have to make the enumerator also public? (The `Enumerable` visibility I understand, as that is needed to avoid boxing.)

Reply:
We need to make it public so that the C# compiler can [duck-typing](https://dev.to/fluffynuts/net-and-duck-typing-part-1-30ce) and use the struct Enumerable directly (use the type in the IL and don't cast to `IEnumerable`). This avoids boxing and any allocation of the enumerable.
If we return an IEnumerator, it's not possible (at least for the .NET Framework) to change this at JIT time.

The same optimisation also happens in the BCL for types like [`List<T>`](https://github.com/dotnet/runtime/blob/1d1bf92fcf43aa6981804dc53c5174445069c9e4/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L665).